Implement IUtf8SpanFormattable on DateTime, DateTimeOffset, DateOnly, TimeOnly, TimeSpan, Char, Rune

stephentoub opened the pull request:
Contributes to https://github.com/dotnet/runtime/issues/81500

(Note that we have a fair amount of code duplication between various formatters (like the DateTimeFormat helper updated in this PR), FormattingHelpers used by Utf8Formatter, and Number.Formatting.cs used by formatting for our primitives.  I've done a little consolidation in this PR, but there's a lot more than can / should be done.  We're also inconsistent as to whether we're working in terms of spans, refs, or pointers, which leads to more duplication.  That can also be consolidated further subsequently.)

stephentoub committed:
Implement IUtf8SpanFormattable on DateTime, DateTimeOffset, DateOnly, TimeOnly, TimeSpan, Char, Rune

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Formatter/FormattingHelpers.cs:
```diff
@@ -92,37 +93,44 @@ public static void WriteDigitsWithGroupSeparator(ulong value, Span<byte> buffer)
             {
-                Number.WriteTwoDigits(bufferPtr + startingIndex, value);
-                Number.WriteTwoDigits(bufferPtr + startingIndex + 2, remainder);
+                if (typeof(TChar) == typeof(char))
+                {
+                    Number.WriteTwoDigits((char*)bufferPtr + startingIndex, value);
+                    Number.WriteTwoDigits((char*)bufferPtr + startingIndex + 2, remainder);
+                }
+                else
+                {
+                    Number.WriteTwoDigits((byte*)bufferPtr + startingIndex, value);
+                    Number.WriteTwoDigits((byte*)bufferPtr + startingIndex + 2, remainder);
+                }
```
Why does this need to be split into the `if/else`?

stephentoub replied:
We have two overloads of WriteTwoDigits, each of which does a single read/write: the one for bytes reads/writes a ushort and the one for chars reads/writes a uint.  So I split the calling path to use them.  Is there a better solution?

tannergooding replied:
Thanks for the explanation!

Pushing the branching/specialization down to `WriteTwoDigits` would help simplify the higher calls a bit more. But I don't have a particular preference.

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Formatter/FormattingHelpers.cs:
```diff
@@ -92,37 +93,44 @@ public static void WriteDigitsWithGroupSeparator(ulong value, Span<byte> buffer)
                 uint temp = '0' + value;
                 value /= 10;
-                buffer[i] = (byte)(temp - (value * 10));
+                buffer[i] = TChar.CreateTruncating(temp - (value * 10));
```
It "shouldn't" be an issue, but we have seen some edge cases where the JIT gives up on inlining `CreateTruncating` due to its size.

Would be good to have some basic perf numbers (or disasm checks) showing nothing has regressed for these changes.

If anything has regressed, there are likely some small JIT side fixes we can do to better account for this and avoid the regression

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Globalization/DateTimeFormat.cs:
```diff
@@ -738,16 +723,54 @@ private static bool IsUseGenitiveForm(ReadOnlySpan<char> format, int index, int
+
+        internal static void FormatFraction<TChar>(ref ValueListBuilder<TChar> result, int fraction, ReadOnlySpan<char> fractionFormat) where TChar : unmanaged, IBinaryInteger<TChar>
+        {
+            // TODO: Update when Int32 implements IUtf8SpanFormattable
```
Could we ensure a corresponding issue is logged for this? Very easy to miss/forget about `TODO` comments otherwise.

tarekgh replied:
Or just throw/assert if we get wrong type.

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Formatter/FormattingHelpers.cs:
```diff
@@ -131,14 +139,21 @@ public static unsafe void WriteFourDecimalDigits(uint value, Span<byte> buffer,
-            fixed (byte* bufferPtr = &MemoryMarshal.GetReference(buffer))
+            fixed (TChar* bufferPtr = &MemoryMarshal.GetReference(buffer))
             {
-                Number.WriteTwoDigits(bufferPtr + startingIndex, value);
```
>Number [](http://example.com/codeflow?start=16&length=6)

do we need to assert the TChar type (to be byte)?

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Collections/Generic/ValueListBuilder.cs:
```diff
@@ -58,13 +58,36 @@ public void Append(T item)
+            _pos += source.Length;
+        }
+
+        public Span<T> AppendSpan(int length)
```
>length [](http://example.com/codeflow?start=38&length=6)

maybe it is wroth asserting the length is greater than 0.

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Globalization/DateTimeFormat.cs:
```diff
@@ -1251,50 +1279,56 @@ internal static bool TryFormatTimeOnlyR(int hour, int minute, int second, Span<c
+            {
+                Span<char> dest = MemoryMarshal.Cast<TChar, char>(destination);
+
+                FormattingHelpers.CopyFourChars("Sun,Mon,Tue,Wed,Thu,Fri,Sat,".AsSpan(4 * (int)dayOfWeek), dest);
```
>dayOfWeek [](http://example.com/codeflow?start=95&length=9)

worth asserting the value of dayOfWeek. I know the original code was not doing it.

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Globalization/TimeSpanFormat.cs:
```diff
@@ -137,24 +144,33 @@ private static bool TryFormatStandard(TimeSpan value, StandardFormat format, str
-                    requiredOutputLength += fractionDigits + 1; // digits plus leading decimal separator
+                    requiredOutputLength += fractionDigits;
+                    requiredOutputLength += typeof(TChar) == typeof(char) ?
+                        decimalSeparator.Length :
```
>decimalSeparator [](http://example.com/codeflow?start=24&length=16)

Thanks for fixing this :-)

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Globalization/TimeSpanFormat.cs:
```diff
@@ -137,24 +144,33 @@ private static bool TryFormatStandard(TimeSpan value, StandardFormat format, str
+                    requiredOutputLength += fractionDigits;
+                    requiredOutputLength += typeof(TChar) == typeof(char) ?
+                        decimalSeparator.Length :
+                        Encoding.UTF8.GetByteCount(decimalSeparator);
```
>Encoding.UTF8.GetByteCount(decimalSeparator); [](http://example.com/codeflow?start=24&length=45)

maybe better initialize the length when initializing decimalSeparator and avoid calling this every time we format?

xtqqczze commented on src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Utf8Formatter/FormattingHelpers.cs:
```diff
@@ -147,6 +162,11 @@ public static unsafe void WriteTwoDecimalDigits(uint value, Span<byte> buffer, i
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void CopyFourChars(ReadOnlySpan<char> source, Span<char> destination) =>
+            Unsafe.WriteUnaligned(ref Unsafe.As<char, byte>(ref MemoryMarshal.GetReference(destination)),
```
```suggestion
            Unsafe.WriteUnaligned(ref Unsafe.As<char, byte>(ref MemoryMarshal.GetReference(destination)),
```
```suggestion
            // TODO: Replace with Unsafe.CopyBlockUnaligned once https://github.com/dotnet/runtime/issues/69220 is completed.
            Unsafe.WriteUnaligned(ref Unsafe.As<char, byte>(ref MemoryMarshal.GetReference(destination)),
```

stephentoub committed:
Merge branch 'main' into datetimeutf8format

stephentoub committed:
Address PR feedback

Also dedup Utf8Formatter for TimeSpan with TimeSpan's new IUtf8SpanFormattable implementation and a little more cleanup.

And fix parameter name of TryFormat to match approved name.

stephentoub commented:
```C#
private DateTime _dt = new DateTime(2023, 04, 06, 10, 37, 12, DateTimeKind.Utc);
private TimeSpan _ts = TimeSpan.FromSeconds(12345.6789);
private char[] _chars = new char[100];
private byte[] _bytes = new byte[100];

[Benchmark] public bool DT_Char_R() => _dt.TryFormat(_chars, out _, "r");
[Benchmark] public bool DT_Char_O() => _dt.TryFormat(_chars, out _, "o");
[Benchmark] public bool DT_Char_Rfc1123() => _dt.TryFormat(_chars, out _, DateTimeFormatInfo.InvariantInfo.RFC1123Pattern);
[Benchmark] public bool TS_Char_C() => _ts.TryFormat(_chars, out _, "c");
[Benchmark] public bool TS_Char_G() => _ts.TryFormat(_chars, out _, "G");

[Benchmark] public bool UTF8Formatter_DT_R() => Utf8Formatter.TryFormat(_dt, _bytes, out _, new StandardFormat('R'));
[Benchmark] public bool UTF8Formatter_DT_O() => Utf8Formatter.TryFormat(_dt, _bytes, out _, new StandardFormat('O'));
[Benchmark] public bool UTF8Formatter_TS_C() => Utf8Formatter.TryFormat(_ts, _bytes, out _, default);
[Benchmark] public bool UTF8Formatter_TS_G() => Utf8Formatter.TryFormat(_ts, _bytes, out _, new StandardFormat('G'));
```

|             Method |         Toolchain |      Mean |    Error |   StdDev | Ratio |
|------------------- |------------------ |----------:|---------:|---------:|------:|
|          DT_Char_R | \main\corerun.exe |  21.79 ns | 0.220 ns | 0.206 ns |  1.00 |
|          DT_Char_R |   \pr\corerun.exe |  19.25 ns | 0.376 ns | 0.352 ns |  0.88 |
|                    |                   |           |          |          |       |
|          DT_Char_O | \main\corerun.exe |  27.67 ns | 0.171 ns | 0.151 ns |  1.00 |
|          DT_Char_O |   \pr\corerun.exe |  24.99 ns | 0.162 ns | 0.144 ns |  0.90 |
|                    |                   |           |          |          |       |
|    DT_Char_Rfc1123 | \main\corerun.exe | 179.71 ns | 1.347 ns | 1.194 ns |  1.00 |
|    DT_Char_Rfc1123 |   \pr\corerun.exe | 175.81 ns | 1.055 ns | 0.881 ns |  0.98 |
|                    |                   |           |          |          |       |
|          TS_Char_C | \main\corerun.exe |  23.47 ns | 0.176 ns | 0.164 ns |  1.00 |
|          TS_Char_C |   \pr\corerun.exe |  21.51 ns | 0.127 ns | 0.112 ns |  0.92 |
|                    |                   |           |          |          |       |
|          TS_Char_G | \main\corerun.exe |  33.72 ns | 0.210 ns | 0.186 ns |  1.00 |
|          TS_Char_G |   \pr\corerun.exe |  32.16 ns | 0.666 ns | 0.712 ns |  0.96 |
|                    |                   |           |          |          |       |
| UTF8Formatter_DT_R | \main\corerun.exe |  18.47 ns | 0.101 ns | 0.095 ns |  1.00 |
| UTF8Formatter_DT_R |   \pr\corerun.exe |  18.60 ns | 0.075 ns | 0.062 ns |  1.01 |
|                    |                   |           |          |          |       |
| UTF8Formatter_DT_O | \main\corerun.exe |  23.76 ns | 0.095 ns | 0.084 ns |  1.00 |
| UTF8Formatter_DT_O |   \pr\corerun.exe |  25.30 ns | 0.084 ns | 0.078 ns |  1.06 |
|                    |                   |           |          |          |       |
| UTF8Formatter_TS_C | \main\corerun.exe |  21.20 ns | 0.130 ns | 0.115 ns |  1.00 |
| UTF8Formatter_TS_C |   \pr\corerun.exe |  19.19 ns | 0.062 ns | 0.058 ns |  0.90 |
|                    |                   |           |          |          |       |
| UTF8Formatter_TS_G | \main\corerun.exe |  22.57 ns | 0.254 ns | 0.212 ns |  1.00 |
| UTF8Formatter_TS_G |   \pr\corerun.exe |  22.44 ns | 0.071 ns | 0.060 ns |  0.99 |

EgorBo commented:
Improvements on arm64:

- https://github.com/dotnet/perf-autofiling-issues/issues/16767
- https://github.com/dotnet/perf-autofiling-issues/issues/16757
